source/extensions/filters/common/rbac/engine_impl.cc

```diff
@@ -80,6 +102,68 @@ bool RoleBasedAccessControlEngineImpl::checkPolicyMatch(
   return matched;
 }
 
+RoleBasedAccessControlMatcherEngineImpl::RoleBasedAccessControlMatcherEngineImpl(
+    const xds::type::matcher::v3::Matcher& matcher,
+    Server::Configuration::ServerFactoryContext& factory_context,
+    ActionValidationVisitor& validation_visitor, const EnforcementMode mode)
+    : mode_(mode) {
+  ActionContext context{false};
+  Envoy::Matcher::MatchTreeFactory<Matching::MatchingData, ActionContext> factory(
+      context, factory_context, validation_visitor);
+  matcher_ = factory.create(matcher)();
+  has_log_ = context.has_log_;
+
+  if (!validation_visitor.errors().empty()) {
+    // TODO(snowp): Output all violations.
+    throw EnvoyException(fmt::format("requirement violation while creating route match tree: {}",
+                                     validation_visitor.errors()[0]));
+  }
+}
+
+bool RoleBasedAccessControlMatcherEngineImpl::handleAction(const Network::Connection& connection,
+                                                           StreamInfo::StreamInfo& info,
+                                                           std::string* effective_policy_id) const {
+  return handleAction(connection, *Http::StaticEmptyHeaders::get().request_headers, info,
+                      effective_policy_id);
+}
+
+bool RoleBasedAccessControlMatcherEngineImpl::handleAction(
+    const Network::Connection& connection, const Envoy::Http::RequestHeaderMap& headers,
+    StreamInfo::StreamInfo& info, std::string* effective_policy_id) const {
+  Matching::MatchingDataImpl data(connection, headers, info);
+  const auto& result = Envoy::Matcher::evaluateMatch<Matching::MatchingData>(*matcher_, data);
+  ASSERT(result.match_state_ == Envoy::Matcher::MatchState::MatchComplete);
+  if (result.result_) {
```

Comment:
I'm not very familiar with the new matcher so please correct me if my understanding is wrong. I'm wondering what does the `result.result_` mean here? Is it true if the matcher is matched and false if not matched? If so what happens if the DENY is allow and the matcher is not matched (this should be allowed but it seems we will still deny in line 164)?

Reply:
> I'm wondering what does the result.result_ mean here? Is it true if the matcher is matched and false if not matched?

`result.result_` tells the matcher what action should be take. In our implementation, the `result.result_` is a `Envoy::Extension::Filters::Common::RBAC::Action`. The `Action` has a field called `action_`, which is the `envoy::config::rbac::v3::RBAC::Action`.

> If so what happens if the DENY is allow and the matcher is not matched

If not matched, `result.result_` will be `nullptr`, and the request falls back to the default DENY.